src/ghga_service_commons/api/api.py

```diff
@@ -125,6 +133,80 @@ class ApiConfigBase(BaseSettings):
             + " are always allowed for CORS requests."
         ),
     )
+    generate_correlation_id: bool = Field(
+        True,
+        examples=[True, False],
+        description=(
+            "A flag, which, if False, will result in an error when inbound requests don't"
+            + " possess a correlation ID. If True, requests without a correlation ID will"
+            + " be assigned a newly generated ID in the correlation ID middleware function."
+        ),
+    )
+
+
+def set_header_correlation_id(request: Request, correlation_id: str):
+    """Set the correlation ID on the request header. Modifies the header in-place."""
+    headers = request.headers.mutablecopy()
+    headers[CORRELATION_ID_HEADER_NAME] = correlation_id
+    request.scope.update(headers=headers.raw)
+    # delete _headers to force update
+    delattr(request, "_headers")
+    log.info("Assigned %s as header correlation ID value.", correlation_id)
+
+
+def get_validated_correlation_id(
+    correlation_id: str, generate_correlation_id: bool
+) -> str:
+    """Returns valid correlation ID.
+
+    If `correlation_id` is valid, it returns that.
+    If it is empty and `generate_correlation_id` is True, a new value is generated.
+    Otherwise, an error is raised.
+
+    Raises:
+        InvalidCorrelationIdError: If a correlation ID is invalid or empty (and
+            `generate_correlation_id` is False).
+    """
+    if not correlation_id and generate_correlation_id:
+        correlation_id = new_correlation_id()
+        log.info("Generated new correlation id: %s", correlation_id)
+    else:
+        validate_correlation_id(correlation_id)
+    return correlation_id
+
+
+async def correlation_id_middleware(
+    request: Request, call_next, generate_correlation_id: bool
+):
+    """Ensure request header has a valid correlation ID.
+
+    Set the correlation ID ContextVar before passing on the request.
+
+    Raises:
+        InvalidCorrelationIdError: If a correlation ID is invalid or empty (and
+            `generate_correlation_id` is False).
+    """
+    correlation_id = request.headers.get(CORRELATION_ID_HEADER_NAME, "")
+
+    # If a correlation ID exists, validate it. If not, generate a new one.
```

Comment:
Should be rephrased as this depends on the boolean value and might be misleading